Deposits

nikhilsaraf opened the pull request:
I added the Deposits tab which reads currencies from the available balances (i.e. trusted set of asset codes). This includes a list of anchors that support deposits for these currencies, not all of which will be trusted.

In order to deposit using an anchor, one would need to add a trust line for the anchor first and then deposit based on the instructions provided.

This PR will only include the deposits component, not withdrawals.

related to https://github.com/irisli/stellarterm/issues/62

- [x] Currency dropdown and anchor list in a new Deposits tab
- [x] Feature flag to hide Deposits tab until it's fully finished

nikhilsaraf commented:
Here's a screenshot of how it looks right now.

![screen shot 2017-10-16 at 5 28 13 pm](https://user-images.githubusercontent.com/1028334/31641375-0bb5cbc0-b299-11e7-8b63-90928f09bb91.png)

irisli commented:
Let me know when it is ready for review.

nikhilsaraf commented:
If you are ok with the screenshot (in particular the "Coming Soon" text since icons is just a UI concern) then I'd say this is ready for review -- I'd rather merge this first to keep it well scoped out. I'll continue to work on the 2 unchecked items in parallel.


> On Oct 16, 2017, at 6:42 PM, Iris Li <notifications@github.com> wrote:
> 
> Let me know when it is ready for review.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

nikhilsaraf commented:
On thinking about this more, if you want I could disable the deposits tab if you don't want to show it for now, but leaving all the code in there --this will prevent code drift / merge conflicts, while working on the other two pieces (I think it will be better to split up these 3 checkbox into separate PRs).

irisli commented:
Okay, the feature flag would be good. Can you put it in a feature flag. I'll look through it tomorrow.

Additionally, I'd like to not link the AssetCards to the issuer websites. I had previously done that but it causes some negative things:
- Easy to accidentally click into the issuer website. (Most of the time, users aren't clicking to get to the issuer website).
- Increases the attack surface for phishers. (We want to users to type in the addresses. And it's trivial to do so)
- Problems when we want to actually link the asset somewhere internally. Having two possible actions for a single button is confusing to the user.

nikhilsaraf commented:
ok, I'll add a feature flag. --how are you doing feature flags, is it just a boolean in the code somewhere?
ok, will remove the links; fyi @jedmccaleb since you had suggested adding the links.

nikhilsaraf commented:
Removed the link to the anchor website.
Added a feature flag in Session.jsx, you can still go to the `/deposit` url directly to see the deposits page and enabling this now is very easy.
I don't have any more planned changes for this PR, I'll move the checkboxes from above to the [original ticket](https://github.com/irisli/stellarterm/issues/62).

irisli commented:
Oh, yeah i don't have any formal feature flags. I think the feature should just be enabled but have no way to get to it unless you know the url.

nikhilsaraf commented:
Cool, that's what the most recent commit does. The URL is /deposit
Let me know if you have any concerns over this PR otherwise it's good on my end
(See the main ticket for the two pending issues)

> On Oct 18, 2017, at 8:40 AM, Iris Li <notifications@github.com> wrote:
> 
> Oh, yeah i don't have any formal feature flags. I think the feature should just be enabled but have no way to get to it unless you know the url.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

irisli commented on src/components/Session/Deposit.jsx:
```diff
@@ -0,0 +1,65 @@
+        d={this.props.d}
+        allAssets={allAssets}
+        selectedAsset={this.state.selectedAsset || allAssets[0].code}
+        onCurrencyChange={this._onCurrencyChange.bind(this)}
```
Simpler and easier to maintain

```js
onCurrencyChange={newAsset => this.setState({'selectedAsset': newAsset})}
```

irisli replied:
Also, note that theres no need for bind since fat arrow automatically does it

nikhilsaraf replied:
ok, will update

irisli commented on src/styles/solar-stellarterm/styles/_forms.base.scss:
```diff
@@ -17,6 +17,13 @@ html input.is-disabled {
   cursor: not-allowed;
 }
+
+.deposit-currency .dropdown {
```
This does not belong in this file. `solar-stellarterm` is a theme.

Instead:
1. Make a `_Deposit.scss`
2. Include it in `app.scss`
3. Rename this class to `DepositCurrency__dropdown`

nikhilsaraf replied:
ok, since I've added a second commit on top of this, is it ok with you if I leave this as-is in this commit, and then make this css change in the second commit which has the _Deposit.scss? that will be a little easier to make the change

irisli commented on src/components/Session/TrustButton.jsx:
```diff
@@ -0,0 +1,61 @@
+    }
+  }
+
+  _handleSubmitTrust(event) {
```
No need for underscores because we never call anything from a react component.

Also, can you put this back into the constructor with the fat arrow `()=>{}` function so that theres less binding happening?

Thanks

nikhilsaraf replied:
ok, I'll move this (idiomatic React code discourages having any data on the `this` object).

nikhilsaraf replied:
also, the fat arrow is implicitly binding under the covers anyways :)

irisli replied:
The implicit binding is nice because it reduces overhead. By just using fat arrows, we almost never have to worry about binding. With having a class function, if someone is tired, they might forget to use `bind` and then have trouble figuring out a bug.

I prefer to have simplicity and ease of coding over following dogma.

nikhilsaraf commented:
Made all the changes you suggested, this is ready to merge. Let me know if you have any questions or concerns.

irisli commented on .gitignore:
```diff
@@ -3,6 +3,7 @@ dist/
 .sass-cache/
 bower_components/
 src/images.js
+src/currencies.js
```
Lets just keep everything inside on src/images.js file

irisli commented on directory/DirectoryBuilder.js:
```diff
@@ -93,6 +93,7 @@ class DirectoryBuilder {
       code: details.code,
       issuer: details.issuer,
       domain: anchorDomain,
+      instructions: details.instructions || "Coming Soon...",
```
This is too opinionated and is promising something is coming soon.

Instructions should only be in an asset if it is set. If not set, then instructions should not appear in the asset object.

irisli commented on gulpfile.js:
```diff
@@ -55,6 +55,36 @@ gulp.task('images', (cb) => {
 });
 
+// Images (For big images that get turned into base64)
+gulp.task('currencies', (cb) => {
```
Lets just piggy back off of the images gulp task to keep things simpler

irisli commented on src/components/Session/Deposit.jsx:
```diff
@@ -0,0 +1,68 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      'selectedAsset': null
```
Quote not necessary here, but you don't have to fix it.

nikhilsaraf commented:
changes from our discussion outside of this PR (removed coming soon text and removed currency images).

irisli commented:
![image](https://user-images.githubusercontent.com/5728307/31802086-0ba40862-b501-11e7-874d-a163f56b8a3e.png)

I fixed some small code details and reimplemented the "feature flagging" (leaving it out of the nav bar).

irisli commented:
The design of this needs some work before we can go live with it
